python/shotgun_model/data_handler_cache.py

```diff
@@ -0,0 +1,198 @@
+# Copyright (c) 2016 Shotgun Software Inc.
+#
+# CONFIDENTIAL AND PROPRIETARY
+#
+# This work is provided "AS IS" and subject to the Shotgun Pipeline Toolkit
+# Source Code License included in this distribution package. See LICENSE.
+# By accessing, using, copying or modifying this work you indicate your
+# agreement to the Shotgun Pipeline Toolkit Source Code License. All rights
+# not expressly granted therein are reserved by Shotgun Software Inc.
+
+
+from .util import compare_shotgun_data
+
+class ShotgunDataHandlerCache(object):
+    """
+    Low level convenience wrapper around a data handler cache.
+    Contains a dictionary structure of simple objects suitable
+    for fast serialization with pickle.
+
+    Used in conjunction with the data handler.
+    """
+
+    # internal constants for serialization performance
+    (CACHE_BY_UID, CACHE_CHILDREN, UID, IS_LEAF, PARENT, FIELD, SG_DATA) = range(7)
+
+    def __init__(self, raw_data=None):
+        """
+        :param raw_data: raw data to initialize with.
+        """
+        if raw_data:
+            self._cache = raw_data
+        else:
+            # init clear cache
+            self._cache = {
+                self.CACHE_CHILDREN: {},  # hierarchy
+                self.CACHE_BY_UID: {},    # uid-based lookup
+                self.UID: None            # the uid of the root is None
+            }
+
+    @property
+    def raw_data(self):
+        """
+        The raw dictionary data contained in the cache.
+        """
+        return self._cache
+
+    @property
+    def size(self):
+        """
+        The number of items in the cache
+        """
+        return len(self._cache[self.CACHE_BY_UID])
+
+    @property
+    def uids(self):
+        """
+        All the uids in this cache as a set
+        """
+        return set(self._cache[self.CACHE_BY_UID].iterkeys())
+
+    def get_child_uids(self, parent_uid):
+        """
+        Returns all the child uids for the given parent
+
+        :param parent_uid: Parent uid
+        :returns: list of child uids
+        """
+        if parent_uid is None:
+            return self._cache[self.CACHE_CHILDREN].keys()
```

Comment:
It would be nice if get_child_uids returned an iterable object (.iterkeys()), just like your other get_multiple_somethings bellow. It would make the interface more consistent. The caller creates a set from that data anyway.